text/2021-02-24-coprocessor-plugin.md

```diff
@@ -0,0 +1,161 @@
+
+### Multi-plugin
+
+Currently TiKV has only one coprocessor `tidb_query`. However, without further work on statically linked plugin and txn mode support, we can't strip it from official release. So, multiple coprocessor has to be supported. Basically, we may need to add a `gPRC` rpc for coprocessor v2 request, in which coprocessor name and version is given, so that TiKV will be able to dispatch the request to the proper coprocessor, as well as to reject the request on version mismatch.
+
+### Protobuf design
+
+```proto
+message RawCoprocessorRequest {
+    kvrpcpb.Context context = 1;
+
+    string copr_name = 2;
+    string copr_version_constraint = 3;
+
+    bytes data = 4;
+}
+
+message RawCoprocessorResponse {
+    bytes data = 1;
+
+    errorpb.Error region_error = 2;
+    string other_error = 3;
+}
+```
+
+### API design
+
+To reduce the learning overhead, it'll be better that the API of the coprocessor plugin get closer to the client. Thus, it'll looks like the `RawClient` in the [Rust Client](https://github.com/tikv/client-rust) with extra txn-like methods e.g. `commit` and `lock`.
+
+```rust
+use std::ops::Range;
+
+pub type Key = Vec<u8>;
+pub type Value = Vec<u8>;
+pub type KvPair = (Key, Value);
+
+#[derive(Debug)]
+pub struct Region {
+    id: u64,
+    start_key: Key,
+    end_key: Key,
+    region_epoch: RegionEpoch,
+}
+
+#[derive(Debug)]
+pub struct RegionEpoch {
+    pub conf_ver: u64,
+    pub version: u64,
+}
+
+#[derive(Debug)]
+pub enum Error {
+    KeyNotInRegion { key: Key, region: Region },
+    // More
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+#[async_trait]
+pub trait RawStorage: Send {
+    async fn get(&self, key: Key) -> Result<Option<Value>>;
+    async fn batch_get(&self, keys: Vec<Key>) -> Result<Vec<KvPair>>;
+    async fn scan(&self, key_range: Range<Key>) -> Result<Vec<Value>>;
+    async fn put(&mut self, key: Key, value: Value) -> Result<()>;
+    async fn batch_put(&mut self, kv_pairs: Vec<KvPair>) -> Result<()>;
+    async fn delete(&mut self, key: Key) -> Result<()>;
+    async fn batch_delete(&mut self, keys: Vec<Key>) -> Result<()>;
+    async fn delete_range(&mut self, key_range: Range<Key>) -> Result<()>;
+}
+
+pub trait Coprocessor: Send + Sync {
+    fn on_raw_coprocessor_request(
+        &self,
+        region: Region,
+        request: Vec<u8>,
+        storage: Box<dyn RawStorage>,
+    ) -> Result<Vec<u8>>;
+```
+
+### Keyspace
+
+Keyspace[[RFC]](https://github.com/tikv/rfcs/pull/39)[[The most updated design doc]](https://docs.google.com/document/d/1x17-urAqToDo8TVXJroEHtc76fdssFaoANjSaNDhjKg/edit) is an incoming feature of TiKV that is highly related to coprocessor plugin. Keyspace determines whether a range of key should only be used in transaction mode or in RawKV mode. Since coprocessor works in either RawKV mode or txn mode, surely coprocessor plugin framework should aware of Keyspace. The details is TBD.
```

Comment:
should we consider the coprocessor plugin after the key range feature is implemented?

Reply:
It would be the best if the keyspace is done. But it's also fine to implement coprocessor plugin in advance since we're assuming the user takes responsibility on the key range seperation.